src/test/scala/org/apache/spark/sql/delta/OptimisticTransactionSuite.scala

```diff
@@ -94,6 +95,51 @@ class OptimisticTransactionSuite extends QueryTest with SharedSQLContext {
     }
   }
 
+  private val addA_P1 = AddFile("a", Map("part" -> "p1"), 1, 1, dataChange = true)
+  private val addB_P1 = AddFile("b", Map("part" -> "p1"), 1, 1, dataChange = true)
+  private val addC_P2 = AddFile("c", Map("part" -> "p2"), 1, 1, dataChange = true)
+  private val addD_P2 = AddFile("d", Map("part" -> "p2"), 1, 1, dataChange = true)
+
+  test("allow modify independent partitions") {
+    withTempDir { tempDir => withSQLConf(ENABLE_CONCURRENT_PARTITIONS_WRITE.key -> "true") {
+      val log = DeltaLog(spark, new Path(tempDir.getCanonicalPath))
+      // Initialize the log and add data. Truncate() is just a no-op placeholder.
+      log.startTransaction().commit(addA_P1 :: addD_P2 :: Nil, Truncate())
+
+      val tx1 = log.startTransaction()
+      tx1.filterFiles()
+
+      // tx2 commits before tx1
+      val tx2 = log.startTransaction()
+      tx2.filterFiles()
+      tx2.commit(addA_P1.remove :: addB_P1 :: Nil, Truncate())
+
+      tx1.commit(addC_P2 :: addD_P2.remove :: Nil, Truncate())
+      checkAnswer(log.update().allFiles.select("path"), Row("b") :: Row("c") :: Nil)
+    }}
+  }
+
+  test("block modify dependent partitions") {
+    withTempDir { tempDir => withSQLConf(ENABLE_CONCURRENT_PARTITIONS_WRITE.key -> "true") {
+      val log = DeltaLog(spark, new Path(tempDir.getCanonicalPath))
+      // Initialize the log and add data. Truncate() is just a no-op placeholder.
+      log.startTransaction().commit(addA_P1 :: addD_P2 :: Nil, Truncate())
+
+      val tx1 = log.startTransaction()
+      tx1.filterFiles()
+
+      // tx2 commits before tx1
+      val tx2 = log.startTransaction()
+      tx2.filterFiles()
+      tx2.commit(addA_P1.remove :: addB_P1 :: Nil, Truncate())
+
+      intercept[ConcurrentWriteException] {
+        // addB_P1 is manipulated in both transactions
+        tx1.commit(addA_P1 :: addB_P1.remove :: Nil, Truncate())
+      }
+    }}
+  }
```

Comment:
There's a lot of test coverage missing here. Off the top of my head, we also need to test (at minimum):

* nested partitions
* different partition columns with the same set of partition values
* multiple transactions with actions other than AddFiles (metadata changes, remove files)
* multiple transactions which may read data in other partitions (or a shared partitions) but only write to a different partition. This is not a common scenario, but still possible with merge.

Reply:
I can add more test coverage, just a few clarifications:
> * nested partitions
by "nested partitions" you mean partitioning on 2 or more columns ?
> * different partition columns with the same set of partition values
can you write an example, maybe you mean p1: "val1", p2: "val1" ?
> * multiple transactions with actions other than AddFiles (metadata changes, remove files)
metadata, protocol and txn changes are handled in different part of code and are not affected by this change, if RemoveFile's partitioning problem (mentioned above) will be solved I can add test for it
> * multiple transactions which may read data in other partitions (or a shared partitions) but only write to a different partition. This is not a common scenario, but still possible with merge.

**This requires broader changes**. This is the reason for constraint "partitions read == partitions write"
Currently there is no partition tracking for reads - only a global `dependsOnFiles` flag. To track partitioning I would have to add partition argument to `filterFiles` method - currently having 7 usages. I'd like to have your opinion on this...